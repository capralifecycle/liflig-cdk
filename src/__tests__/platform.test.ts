import "@aws-cdk/assert/jest"
import { App, Stack, Construct } from "@aws-cdk/core"
import * as sns from "@aws-cdk/aws-sns"
import * as ssm from "@aws-cdk/aws-ssm"
import "jest-cdk-snapshot"
import {
  PlatformProducer,
  PlatformConsumer,
  PlatformConsumerProps,
  PlatformProducerProps,
} from "../platform"

interface ProducerProps extends PlatformProducerProps {
  platformName: string
  paramNamespace: string
  alarmTopic: sns.Topic
}

const alarmTopicArn = "alarm-topic-arn"

class ExamplePlatformProducer extends PlatformProducer {
  constructor(scope: Construct, id: string, props: ProducerProps) {
    super(scope, id, props)
    this.putParam(alarmTopicArn, props.alarmTopic.topicArn)
  }
}

interface ConsumerProps extends PlatformConsumerProps {
  platformName: string
  paramNamespace: string
}

class ExamplePlatformConsumer extends PlatformConsumer {
  constructor(scope: Construct, id: string, props: ConsumerProps) {
    super(scope, id, props)
  }

  public get alarmTopic(): sns.ITopic {
    return this._alarmTopic()
  }
  private _alarmTopic = this.lazy(() =>
    sns.Topic.fromTopicArn(this, "TopicArn", this.getParam(alarmTopicArn)),
  )
}

test("produce example plaform", () => {
  const app = new App()
  const stack = new Stack(app, "ProducerStack")

  const alarmTopic = new sns.Topic(stack, "Topic", {
    topicName: "alarmtopic",
    displayName: "Alarm topic",
  })

  new ExamplePlatformProducer(stack, "PlatformProducer", {
    platformName: "test",
    paramNamespace: "test",
    alarmTopic: alarmTopic,
  })

  expect(stack).toMatchCdkSnapshot()
})

test("consume example platform", () => {
  const app = new App()
  const stack = new Stack(app, "ConsumerStack")

  const platform = new ExamplePlatformConsumer(stack, "PlatformConsumer", {
    platformName: "test",
    paramNamespace: "test",
  })

  // only here to get the value of topicArn, somehow
  new ssm.StringParameter(stack, "ExampleParam", {
    stringValue: platform.alarmTopic.topicArn,
  })

  expect(stack).toMatchCdkSnapshot()
})
