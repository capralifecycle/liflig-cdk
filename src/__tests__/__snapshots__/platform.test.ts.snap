// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`consume example platform 1`] = `
Object {
  "Parameters": Object {
    "SsmParameterValuelifligcdkplatformNamespaceTestplatformplatformNameTestalarmtopicarnC96584B6F00A464EAD1953AFF4B05118Parameter": Object {
      "Default": "/liflig-cdk/platformNamespaceTest/platform/platformNameTest/alarm-topic-arn",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "ExampleParam05F8FCAC": Object {
      "Properties": Object {
        "Type": "String",
        "Value": Object {
          "Ref": "SsmParameterValuelifligcdkplatformNamespaceTestplatformplatformNameTestalarmtopicarnC96584B6F00A464EAD1953AFF4B05118Parameter",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
}
`;

exports[`produce example plaform 1`] = `
Object {
  "Resources": Object {
    "PlatformProduceralarmtopicarnC48B2B79": Object {
      "Properties": Object {
        "Name": "/liflig-cdk/platformNamespaceTest/platform/platformNameTest/alarm-topic-arn",
        "Type": "String",
        "Value": Object {
          "Ref": "TopicBFC7AF6E",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "TopicBFC7AF6E": Object {
      "Properties": Object {
        "DisplayName": "Alarm topic",
        "TopicName": "alarmtopic",
      },
      "Type": "AWS::SNS::Topic",
    },
  },
}
`;
